build_tools/benchmarks/set_adreno_gpu_scaling_policy.sh

```diff
@@ -0,0 +1,80 @@
+#!/bin/sh
+
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Runs on a phone with Adreno GPU to set the GPU frequency scaling policy.
+
+################################### WARNING ####################################
+# This will overheat the phone if it's not on a cooling plate, resulting in    #
+# thermal throttling. To prevent anything catching on fire, the actual GPU     #
+# frequencies will be throttled to below the maximum, skewing your results.    #
+################################################################################
+
+set -euo pipefail
+
+POLICY="${1:-performance}"
+
+readonly ADRENO_GPU_PATH="/sys/class/kgsl/kgsl-3d0"
+
+# Available frequencies are ordered from highest to lowest.
+readonly ADRENO_MAX_FREQ=$(cat "${ADRENO_GPU_PATH}/devfreq/available_frequencies" | tr " " "\n" | head -1)
+readonly ADRENO_MIN_FREQ=$(cat "${ADRENO_GPU_PATH}/devfreq/available_frequencies" | tr " " "\n" | tail -1)
+
+# Power levels match available freqencies.
+readonly ADRENO_MAX_PWRLEVEL=0
+readonly ADRENO_MIN_PWRLEVEL=$(expr $(cat "${ADRENO_GPU_PATH}/num_pwrlevels") - 1)
+
+echo "GPU info (before changing power policy):"
+echo 'model\t\tcur\t\tmin\t\tmax'
+echo "---------------------------------------------------------"
+paste \
+  "${ADRENO_GPU_PATH}/gpu_model" \
+  "${ADRENO_GPU_PATH}/devfreq/cur_freq" \
+  "${ADRENO_GPU_PATH}/devfreq/min_freq" \
+  "${ADRENO_GPU_PATH}/devfreq/max_freq"
+
+echo "Setting GPU power policy to ${POLICY}"
+
+if [[ "$POLICY" == "performance" ]]; then
+  echo 1 > "${ADRENO_GPU_PATH}/force_clk_on"
+  echo 3600000 > "${ADRENO_GPU_PATH}/idle_timer"
+
+  # Some devices only expose the msm-adreno-tz governor, so allow the
+  # following to fail.
+  echo performance > "${ADRENO_GPU_PATH}/devfreq/governor" || true
+
+  echo ${ADRENO_MAX_FREQ} > "${ADRENO_GPU_PATH}/gpuclk"
+  echo ${ADRENO_MAX_FREQ} > "${ADRENO_GPU_PATH}/devfreq/max_freq"
+  echo ${ADRENO_MAX_FREQ} > "${ADRENO_GPU_PATH}/devfreq/min_freq"
+
+  echo ${ADRENO_MAX_PWRLEVEL} > "${ADRENO_GPU_PATH}/max_pwrlevel"
+  echo ${ADRENO_MAX_PWRLEVEL} > "${ADRENO_GPU_PATH}/min_pwrlevel"
+elif [[ "$POLICY" == "ondemand" ]]; then
```

Comment:
Welcome to the land of mobile GPUs. There are no "standard" names for these AFAIK. Each GPU vendor / device manufacturer can feel free to implement its own scaling policy. But it doesn't mean we will have a script for each phone we need to benchmark either. I'm hoping this one could be used for all Adreno GPUs. (Though again different driver versions may affect what knobs are available, etc. It's... complicated.)

The common names allow us to have one single entry in the main script and then dispatch to concrete per-GPU/per-phone scripts to use specific ones. I like to hide the details in the script rather than exposing all of that in the python script.

Reply:
Changed to "performance" vs "default". I guess these should be relatively "standard" ways.